SystemManager/Src/SM_States.cpp

```diff
@@ -0,0 +1,243 @@
+    // vTaskDelete(system_manager->TM_handle);
+    // xTaskCreate(system_manager->TMOperationTask, "TM Thread", &system_manager->TM_handle);
+
+    // system_manager->SM_to_PM_queue = osMessageQueueNew(1, 256, NULL);
+    // system_manager->PM_to_AM_queue = osMessageQueueNew(1, 256, NULL);
+
+    // JUST for the Nov 27th deadline flight:
+
+    system_manager->SM_to_AM_queue = osMessageQueueNew(1U, 256U, NULL);
+    system_manager->attitude_manager.setSmQueue(system_manager->SM_to_AM_queue);
+    AM::AttitudeManager* am_pointer = &system_manager->attitude_manager;
+    xTaskCreate(system_manager->AMOperationTask, "AM Thread", 400U,
+                (void *)am_pointer, osPriorityNormal,
+                &system_manager->AM_handle);
+    // Task will automatically be added to the FreeRTOS scheduler
+}
+
+void FlightMode::execute(SystemManager *system_manager) {
+    // Get sensor fusion data from LOS interfaces
+    // TODO
+    // SM.sf_data = los_interface.
+
+    // Get RC data from LOS
+    system_manager->rc_data = Los_Link::getInstance().getRx();
+    if (system_manager->rc_data.rx_channels[SM::RC_ARM_CHANNEL] < 10) {
+        system_manager->setState(DisarmMode::getInstance());
+    }
+
+    // Decode RC data and convert to AM message type
+    AM::AttitudeManagerInput to_am_data = RcToAmInput(system_manager->rc_data);
+    void* msg_pointer = &to_am_data;
+
+    // Send to AM mail queue (JUST Nov 27th implementation)
+    osMessageQueuePut(system_manager->SM_to_AM_queue, msg_pointer, osPriorityNormal, 0);
+
+    // Get big packet message from TM and store it
+    // TODO
+
+    // Determine flight state (Takeoff/cruise/Landing/landed) and update state
+    // TODO
+    // if ((system_manager->operation_mode == SM::LANDING && landedCheck()) ||
+    // system_manager->tm_packet.state == "SM::DISARMED") {
+    //     system_manager->operation_mode = SM::DISARMED;
+    // }
+
+    // // TODO update this when I get what TM sends/have a landing check
+    // if (system_manager->tm_packet.desired_state == "FATAL_FAIL") {
+    //     // If fatal Failure then move to fatal fail mode
+    //     system_manager->setState(FatalFailMode::getInstance());
+    // } else if (system_manager->operation_mode == SM::DISARMED) {
+    //     // If disarm/landed then move to disarm
+    //     system_manager->setState(DisarmMode::getInstance());
+    // }
+
+    // Get waypoint(s) from TM/RC (through LOS) and pass SF data and waypoint(s)
+    // to PM (who sends to AM)
+    // TODO
+
+    // Get latest message back from AM
+    // TODO
+
+    // Create to-ground packet with AM response and SF and send through mail queue to TM
+    // TODO
+}
+
+void FlightMode::exit(SystemManager *system_manager) {
+    // Stop the AM, PM thread and decrease TM speed
+
+    // Stop PM
+    // vTaskDelete(system_manager->PM_handle);
+    // Stop AM
+    vTaskDelete(system_manager->AM_handle);
```

Comment:
What happens if we delete the task in the middle of AM operation. Is that a concern?

Reply:
Not really a concern. FreeRTOS is just scheduling these and has I believe an idle task that will remove it from the queue. This makes sure that the one resource it can hold (that global flag that isn't in here yet) won't get locked out.